File: forms/audio-vestibular-assessment/back-end-with-loco/tests/engine/audio_vestibular_grader_test.rs
use audio_vestibular_assessment_loco_crate::engine::audio_vestibular_grader::{
    calculate_dhi, calculate_pure_tone_audiometry, grade,
};
use audio_vestibular_assessment_loco_crate::engine::audio_vestibular_rules::{
    calculate_pta_from_thresholds, classify_dhi_handicap, classify_hearing_loss_grade,
    dhi_answer_score,
};
use audio_vestibular_assessment_loco_crate::engine::types::*;

fn ear(hz500: f64, hz1000: f64, hz2000: f64, hz4000: f64) -> EarThresholds {
    EarThresholds {
        hz500: Some(hz500),
        hz1000: Some(hz1000),
        hz2000: Some(hz2000),
        hz4000: Some(hz4000),
    }
}

#[test]
fn empty_assessment_is_unknown_grade() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.hearing_loss_grade, "unknown");
    assert_eq!(result.dhi_total, 0);
    assert_eq!(result.dhi_handicap_level, "no-handicap");
    assert!(result.right_pta.is_none());
    assert!(result.left_pta.is_none());
    assert!(result.better_ear_pta.is_none());
}

#[test]
fn pta_from_thresholds_averages_four_freqs() {
    let pta = calculate_pta_from_thresholds(&ear(10.0, 20.0, 30.0, 40.0));
    assert_eq!(pta, Some(25.0));
}

#[test]
fn pta_with_missing_thresholds_returns_none() {
    let pta = calculate_pta_from_thresholds(&EarThresholds::default());
    assert!(pta.is_none());
}

#[test]
fn classify_grade_boundaries() {
    assert_eq!(classify_hearing_loss_grade(Some(15.0)), "normal");
    assert_eq!(classify_hearing_loss_grade(Some(20.0)), "mild");
    assert_eq!(classify_hearing_loss_grade(Some(34.9)), "mild");
    assert_eq!(classify_hearing_loss_grade(Some(35.0)), "moderate");
    assert_eq!(classify_hearing_loss_grade(Some(50.0)), "moderately-severe");
    assert_eq!(classify_hearing_loss_grade(Some(65.0)), "severe");
    assert_eq!(classify_hearing_loss_grade(Some(80.0)), "profound");
    assert_eq!(classify_hearing_loss_grade(None), "unknown");
}

#[test]
fn pure_tone_calculation_picks_better_ear() {
    let mut data = AssessmentData::default();
    data.pure_tone_audiometry.right_ear.air_conduction = ear(40.0, 40.0, 40.0, 40.0);
    data.pure_tone_audiometry.left_ear.air_conduction = ear(20.0, 20.0, 20.0, 20.0);
    let pta = calculate_pure_tone_audiometry(&data);
    assert_eq!(pta.right_pta, Some(40.0));
    assert_eq!(pta.left_pta, Some(20.0));
    assert_eq!(pta.better_ear_pta, Some(20.0));
    assert_eq!(pta.asymmetry, Some(20.0));
    assert_eq!(pta.hearing_loss_grade, "mild");
}

#[test]
fn dhi_answer_score_values() {
    assert_eq!(dhi_answer_score("yes"), 4);
    assert_eq!(dhi_answer_score("sometimes"), 2);
    assert_eq!(dhi_answer_score("no"), 0);
    assert_eq!(dhi_answer_score(""), 0);
}

#[test]
fn classify_dhi_handicap_boundaries() {
    assert_eq!(classify_dhi_handicap(0), "no-handicap");
    assert_eq!(classify_dhi_handicap(16), "no-handicap");
    assert_eq!(classify_dhi_handicap(18), "mild");
    assert_eq!(classify_dhi_handicap(36), "mild");
    assert_eq!(classify_dhi_handicap(38), "moderate");
    assert_eq!(classify_dhi_handicap(52), "moderate");
    assert_eq!(classify_dhi_handicap(54), "severe");
    assert_eq!(classify_dhi_handicap(100), "severe");
}

#[test]
fn dhi_all_yes_is_maximum() {
    let mut data = AssessmentData::default();
    let dhi = &mut data.dizziness_handicap_inventory;
    dhi.q1 = "yes".to_string();
    dhi.q2 = "yes".to_string();
    dhi.q3 = "yes".to_string();
    dhi.q4 = "yes".to_string();
    dhi.q5 = "yes".to_string();
    dhi.q6 = "yes".to_string();
    dhi.q7 = "yes".to_string();
    dhi.q8 = "yes".to_string();
    dhi.q9 = "yes".to_string();
    dhi.q10 = "yes".to_string();
    dhi.q11 = "yes".to_string();
    dhi.q12 = "yes".to_string();
    dhi.q13 = "yes".to_string();
    dhi.q14 = "yes".to_string();
    dhi.q15 = "yes".to_string();
    dhi.q16 = "yes".to_string();
    dhi.q17 = "yes".to_string();
    dhi.q18 = "yes".to_string();
    dhi.q19 = "yes".to_string();
    dhi.q20 = "yes".to_string();
    dhi.q21 = "yes".to_string();
    dhi.q22 = "yes".to_string();
    dhi.q23 = "yes".to_string();
    dhi.q24 = "yes".to_string();
    dhi.q25 = "yes".to_string();
    let res = calculate_dhi(&data);
    assert_eq!(res.total, 100);
    assert_eq!(res.answered_count, 25);
    assert_eq!(res.handicap_level, "severe");
    // 7 functional × 4 = 28
    assert_eq!(res.functional, 28);
    // 9 emotional × 4 = 36
    assert_eq!(res.emotional, 36);
    // 9 physical × 4 = 36
    assert_eq!(res.physical, 36);
}

#[test]
fn grade_combines_pta_and_dhi() {
    let mut data = AssessmentData::default();
    data.pure_tone_audiometry.right_ear.air_conduction = ear(70.0, 75.0, 80.0, 85.0);
    data.pure_tone_audiometry.left_ear.air_conduction = ear(70.0, 75.0, 80.0, 85.0);
    let dhi = &mut data.dizziness_handicap_inventory;
    dhi.q1 = "yes".to_string();
    dhi.q11 = "yes".to_string();
    let res = grade(&data);
    assert_eq!(res.hearing_loss_grade, "severe");
    assert_eq!(res.dhi_total, 8);
    assert_eq!(res.dhi_handicap_level, "no-handicap");
}

File: forms/audio-vestibular-assessment/front-end-with-svelte/src/lib/engine/audio-vestibular-grader.test.ts
import { describe, it, expect } from 'vitest';
import { grade, calculatePureToneAudiometry, calculateDhi } from './audio-vestibular-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { DHI_ITEMS } from './rules';
import type { AssessmentData, DizzinessHandicapInventory } from './types';

function emptyDhi(): DizzinessHandicapInventory {
	const answers: DizzinessHandicapInventory = {};
	for (const item of DHI_ITEMS) answers['q' + item.num] = '';
	return answers;
}

function createBlankAssessment(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', assessmentDate: '' },
		presentingSymptoms: {
			hearingLoss: '',
			hearingLossSide: '',
			hearingLossOnset: '',
			hearingLossDurationMonths: null,
			tinnitus: '',
			tinnitusSide: '',
			otalgia: '',
			otorrhea: '',
			auralFullness: '',
			vertigo: '',
			vertigoCharacter: '',
			vertigoEpisodeDurationSeconds: null,
			vertigoFrequencyPerWeek: null,
			imbalance: '',
			falls: '',
			fallsLastYearCount: null,
			headacheMigraine: '',
			neurologicalSymptoms: '',
			otherSymptoms: ''
		},
		otoscopicExamination: {
			rightEar: { canalStatus: '', tympanicMembrane: '' },
			leftEar: { canalStatus: '', tympanicMembrane: '' },
			notes: ''
		},
		pureToneAudiometry: {
			rightEar: {
				airConduction: { hz500: null, hz1000: null, hz2000: null, hz4000: null },
				boneConduction: { hz500: null, hz1000: null, hz2000: null, hz4000: null },
				pureToneAverage: null
			},
			leftEar: {
				airConduction: { hz500: null, hz1000: null, hz2000: null, hz4000: null },
				boneConduction: { hz500: null, hz1000: null, hz2000: null, hz4000: null },
				pureToneAverage: null
			},
			betterEarPureToneAverage: null,
			asymmetryDb: null,
			audiometryNotes: ''
		},
		speechAudiometry: {
			rightSrtDb: null,
			leftSrtDb: null,
			rightWordRecognitionPercent: null,
			leftWordRecognitionPercent: null,
			speechAudiometryNotes: ''
		},
		tympanometryAcousticReflexes: {
			rightTympanogram: '',
			leftTympanogram: '',
			rightAcousticReflexes: '',
			leftAcousticReflexes: '',
			notes: ''
		},
		vestibularScreening: {
			headImpulseTest: '',
			dixHallpike: '',
			rombergTest: '',
			tandemGait: '',
			nystagmus: '',
			fukudaSteppingTest: '',
			notes: ''
		},
		dizzinessHandicapInventory: emptyDhi(),
		clinicalImpressionReferral: {
			provisionalDiagnosis: '',
			hearingAidCandidate: '',
			vestibularRehabIndicated: '',
			ent_referral: '',
			neurologyReferral: '',
			imagingRequested: '',
			followUpWeeks: null,
			additionalNotes: ''
		}
	};
}

describe('Pure-tone audiometry grading', () => {
	it('returns unknown grade and null PTA for a blank assessment', () => {
		const r = calculatePureToneAudiometry(createBlankAssessment());
		expect(r.betterEarPta).toBeNull();
		expect(r.asymmetry).toBeNull();
		expect(r.hearingLossGrade).toBe('unknown');
	});

	it('computes the 4-frequency PTA and classifies normal hearing', () => {
		const d = createBlankAssessment();
		d.pureToneAudiometry.rightEar.airConduction = { hz500: 10, hz1000: 10, hz2000: 15, hz4000: 15 };
		d.pureToneAudiometry.leftEar.airConduction = { hz500: 10, hz1000: 10, hz2000: 10, hz4000: 10 };
		const r = calculatePureToneAudiometry(d);
		expect(r.rightPta).toBe(12.5);
		expect(r.leftPta).toBe(10);
		expect(r.betterEarPta).toBe(10);
		expect(r.hearingLossGrade).toBe('normal');
	});

	it('classifies profound hearing loss in the worse ear and grades on the better ear', () => {
		const d = createBlankAssessment();
		d.pureToneAudiometry.rightEar.airConduction = { hz500: 90, hz1000: 95, hz2000: 100, hz4000: 100 };
		d.pureToneAudiometry.leftEar.airConduction = { hz500: 30, hz1000: 30, hz2000: 30, hz4000: 30 };
		const r = calculatePureToneAudiometry(d);
		expect(r.rightHearingLossGrade).toBe('profound');
		expect(r.leftHearingLossGrade).toBe('mild');
		expect(r.betterEarPta).toBe(30);
		expect(r.hearingLossGrade).toBe('mild');
	});

	it('computes inter-aural asymmetry', () => {
		const d = createBlankAssessment();
		d.pureToneAudiometry.rightEar.airConduction = { hz500: 60, hz1000: 60, hz2000: 60, hz4000: 60 };
		d.pureToneAudiometry.leftEar.airConduction = { hz500: 20, hz1000: 20, hz2000: 20, hz4000: 20 };
		const r = calculatePureToneAudiometry(d);
		expect(r.asymmetry).toBe(40);
	});
});

describe('Dizziness Handicap Inventory scoring', () => {
	it('scores 0 for a blank inventory', () => {
		const r = calculateDhi(createBlankAssessment());
		expect(r.total).toBe(0);
		expect(r.answeredCount).toBe(0);
		expect(r.handicapLevel).toBe('no-handicap');
	});

	it('scores yes=4, sometimes=2, no=0 and classifies severe', () => {
		const d = createBlankAssessment();
		for (const item of DHI_ITEMS) d.dizzinessHandicapInventory['q' + item.num] = 'yes';
		const r = calculateDhi(d);
		expect(r.total).toBe(100);
		expect(r.answeredCount).toBe(25);
		expect(r.handicapLevel).toBe('severe');
	});

	it('accumulates per-subscale subtotals', () => {
		const d = createBlankAssessment();
		for (const item of DHI_ITEMS) d.dizzinessHandicapInventory['q' + item.num] = 'sometimes';
		const r = calculateDhi(d);
		// 9 functional, 9 emotional, 7 physical items @ 2 pts each
		expect(r.functional).toBe(18);
		expect(r.emotional).toBe(18);
		expect(r.physical).toBe(14);
		expect(r.total).toBe(50);
		expect(r.handicapLevel).toBe('moderate');
	});
});

describe('Flagged issues detection', () => {
	it('raises no flags for a blank assessment', () => {
		const data = createBlankAssessment();
		const result = grade(data);
		expect(result.additionalFlags).toHaveLength(0);
	});

	it('flags sudden hearing loss as urgent', () => {
		const data = createBlankAssessment();
		data.presentingSymptoms.hearingLoss = 'yes';
		data.presentingSymptoms.hearingLossOnset = 'sudden';
		const result = grade(data);
		const flag = result.additionalFlags.find((f) => f.id === 'FLAG-RED-001');
		expect(flag).toBeTruthy();
		expect(flag?.priority).toBe('urgent');
	});

	it('flags asymmetric hearing loss for retrocochlear screen', () => {
		const data = createBlankAssessment();
		data.pureToneAudiometry.rightEar.airConduction = { hz500: 60, hz1000: 60, hz2000: 60, hz4000: 60 };
		data.pureToneAudiometry.leftEar.airConduction = { hz500: 20, hz1000: 20, hz2000: 20, hz4000: 20 };
		const result = grade(data);
		expect(result.additionalFlags.some((f) => f.id === 'FLAG-PTA-001')).toBe(true);
	});

	it('flags positive Dix-Hallpike (BPPV)', () => {
		const data = createBlankAssessment();
		data.vestibularScreening.dixHallpike = 'positive-right';
		const result = grade(data);
		expect(result.additionalFlags.some((f) => f.id === 'FLAG-VEST-001')).toBe(true);
	});

	it('flags recurrent falls', () => {
		const data = createBlankAssessment();
		data.presentingSymptoms.falls = 'yes';
		data.presentingSymptoms.fallsLastYearCount = 3;
		const flags = detectAdditionalFlags(data, grade(data));
		expect(flags.some((f) => f.id === 'FLAG-FALL-001')).toBe(true);
	});

	it('sorts flags by priority (urgent first)', () => {
		const data = createBlankAssessment();
		data.presentingSymptoms.neurologicalSymptoms = 'yes'; // urgent
		data.presentingSymptoms.tinnitus = 'yes'; // low
		data.vestibularScreening.rombergTest = 'abnormal'; // medium
		const flags = grade(data).additionalFlags;
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});

	it('produces unique flag IDs', () => {
		const data = createBlankAssessment();
		data.presentingSymptoms.hearingLoss = 'yes';
		data.presentingSymptoms.hearingLossOnset = 'sudden';
		data.presentingSymptoms.neurologicalSymptoms = 'yes';
		data.vestibularScreening.dixHallpike = 'positive-left';
		const ids = grade(data).additionalFlags.map((f) => f.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});
